### DISCUSSION TITLE
Replace the tokenizer using configs and spacy init

### DISCUSSION THREAD

**Discussion by dakinggg** [explosion/spaCy#6632] [Help: Coding & Implementations] [answered]
Hi, I just started messing around with spacy projects and configs and 3.0 for scispacy. For now, I'm just trying to use `nlp.after_creation` to replace the tokenizer object. My attempt is to call `spacy init nlp configs/base_sm.cfg output/en_core_sci_sm_base --code scripts/base_project_code.py` where `base_project_code.py` contains
```
import spacy

from scispacy.custom_tokenizer import combined_rule_tokenizer


@spacy.registry.callbacks("replace_tokenizer")
def create_callback():
    def replace_tokenizer(nlp):
        nlp.tokenizer = combined_rule_tokenizer(nlp)
        return nlp

    return replace_tokenizer
```

and in my config I have

```
[nlp]
lang = "en"
pipeline = ["tok2vec","tagger","parser","ner"]
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
disabled = []
before_creation = null
after_pipeline_creation = null

[nlp.after_creation]
@callbacks = "replace_tokenizer"
```

When I run the init command, it succeeds and writes the nlp object to disk, but then when i try to load with `spacy.load`, I get `RegistryError: Cant't find 'replace_tokenizer' in registry spacy -> callbacks. Available names: none`. Could you help correct whatever misunderstanding I have? Thanks!

FWIW, on spacy 2.x, we would just set `nlp.tokenizer = combined_rule_tokenizer(nlp)` and then `nlp.to_disk(...)`

**Comment by adrianeboyd** [comment, marked as answer]
This is a good question, and after looking at all the details, it may be something that needs improvement before the final v3 release because the current options within the config are not ideal for cases where you're extending the tokenizer settings from the language defaults in what should be a one-time initialization step.

The immediate problem above is that your callback needs to be registered before you run `spacy.load()` every time. The code from `spacy train --code` is not included as part of what's saved with `nlp.to_disk()`. If you have registered functions that you need for your config, it should be easier from v3.0.0rc3 on (not yet published!) to include this code in your model package with `spacy package` so any registered functions can be imported automatically before loading the model without extra steps on the user's part. (See https://github.com/explosion/spaCy/pull/6531.) If you're running `spacy.load()` from a local path rather than a package, you'll still have to make sure your script imports the code with the registered functions in advance.

Instead of `after_creation`, it would be a bit faster to just modify the language defaults in `before_creation` so that you're not creating the tokenizer twice.

It also seems non-ideal that your callback is going to run every time the model is loaded even though it really only needs to be run once on initialization. It's pretty confusing that when the model is loaded from disk, the callback will run and then the tokenizer settings loaded from the saved model will clobber all those changes. It's not necessarily clear to someone reading the config where the tokenizer settings are coming from at this point. The other language defaults aren't saved in the model, so having this callback makes sense for customizing them (aside from minor variations in the settings across patch versions, where I think it might make sense to version the whole set of defaults, e.g. `spacy.EnglishDefaults.v25`, but there's a lot of overhead involved in doing this).

You could potentially remove the callback from your config after initialization so that when the model is saved the first time there are no traces of the original callback. Then you don't need the registered function after the training step and the tokenizer loads its settings as expected, but this is something we'd really rather discourage because it obscures things rather than making the pipeline configuration easier to reproduce.

This exact kind of problem is why the `[initialize]` block was added and I feel like these kinds of modifications would be a better fit for an `[initialize.tokenizer]` block, but we'll have to discuss it a bit to see what makes sense.

**Comment by dakinggg** [comment]
Thanks for the detailed answer! I definitely understand some things a bit better. From my perspective, being able to set the tokenizer to a custom `Tokenizer` object inside of [initialize.tokenizer], in a way that doesn't require a callback to be run every time the model is loaded, would be ideal I think. That would 1) allow me to replace the tokenizer 2) provide self documentation in the config file 3) not require any callbacks. For now, it seems like the suggested way to do this would be to change the english default tokenizer via `before_creation`, and then include the code with that callback using spacy package/importing the file when testing stuff out?

**Comment by adrianeboyd** [comment]
Yes, for now you can use `before_creation` (or `after_creation`, it's not a huge difference) and import the callback whenever you load the model. If you want to compile v3 (the `develop` branch) from source, you can also try out the `spacy package` updates.

**Comment by dakinggg** [comment]
@adrianeboyd I think the `spacy evaluate` command also needs ` --code` arg?

**Reply by dakinggg** [reply]
pulling this out into its own discussion, since I accepted an answer to my original question

**Comment by adrianeboyd** [comment]
We decided that top-level initialize block callbacks made the most sense, so v3.0.0rc3 should add `initialize.before_init` and `initialize.after_init` callbacks, see #6715. There are examples added to the docs in that PR for customizing the tokenizer settings.

Tokenizer modifications make the most sense in `initialize.before_init` so the tokenizer is configured correctly before the pipeline components are initialized. There shouldn't really be any interactions because `initialize` should use reference `Example` docs and not pay attention to the predicted tokenization (misalignments are handled in `get_loss`), but just in case there are interactions that I'm not envisioning, I think `before_init` is the best recommendation.
